### ISSUE TITLE
[M2.5, ECO, I2C] Restart condition violates I2C Specification

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#18721]
**Problem**
A RESTART condition in I2C is signalled by a START condition, where the SDA goes low whilst SCL is high, occurring after a previous transaction has finished without signaling a STOP condition. To achieve this the SDA line must first go high (it will be low following the ACK bit). In our I2C block, the SDA line will go high in the same clock cycle as the SCL line goes high when it is producing a restart. We believe this is in violation of the I2C specification as documented by NXP, which states:

> The data on the SDA line must be stable during the HIGH period of the clock. The HIGH or LOW state of the data line can only change when the clock signal on the SCL line is LOW

The START and STOP conditions are a clear violation of this, but obviously allowable ones. There is no explicit exception for the preparation for a RESTART so the current behavior can be seen as a violation.

It is unclear how much of an issue this is in practice, the I2C block has been observed successfully working with a number of real I2C devices connected to FPGA whilst employing this violating RESTART behavior.

With the violating RESTART receivers might interpret it as a RESTART or as a STOP followed by an immediate START. Some devices use a RESTART as part of their access protocol where a RESTART indicates a continuation and a STOP indicates an end. Here a RESTART interpreted as a STOP followed by a START would prevent full usage of the device. In some instances the use of RESTART can be avoided by employing higher transaction overhead (e.g. doing many separate small operations rather than a single larger one).

It is notable that whilst there are setup and hold times for the START and STOP conditions specified, there is no specified constraint on how close to SCL rising SDA can rise for this specific scenario (which is a preparation for a START after an ACK, not a START or STOP itself). The constraint that is specified is data setup time (time before SCL positive edge data bit must be stable on SDA, tSU;DAT on the diagram below) but this seemingly only applies to the 8 data bits, with the 9th bit being labelled in the timing diagram but with no matching setup or hold times. Still it seems reasonable to assume that tSU;DAT setup time should be adhered to here also.

Practically violating this setup time may not pose a problem in practice. Internal circuitry relating to processing START/STOP/RESTART in the device must be able to sample SDA whilst SCL remains high (to detect START and STOP) so cannot be clocked directly from SCL so we shouldn’t get metastability issues from SDA changing too closely to SCL. Rather some higher frequency clock will be required to sample SDA to look for the edge, which must have a synchroniser on it as it can’t be guaranteed to have edges synchronous to SCL. Here that circuitry will sample SDA some time after the SCL positive edge and see it is asserted, then sample again and see that it has dropped and see a RESTART.

This is also tracked as a separate issue: #18718 (for tracking it is an I2C bug, this issue is for ECO discussion and impact assessment).

![i2c-timing](https://github.com/lowRISC/opentitan/assets/471032/707c2db3-8529-47b2-b678-8b7984b32489)

**Solutions**

1. Do not employ the RESTART condition. This is entirely under the control of software. However this may result in less efficient communication with some devices and entirely prevent accessing functionality of others.

2. Apply an ECO against the I2C state machine. The state machine controls the SDA and SCL lines and sets them using combinational logic factoring in the state and various other signals. The SDA and SCL outputs are flopped before they are sent to the pads. It is a significant state machine so the synthesized logic will be complex. The SDA line needs to be forced high when in a certain state and the ‘pend_restart’ signal is set. It is possible this could be achieved with a couple of extra gates depending on what is available in the synthesised logic. If there is a signal asserted when in the appropriate state and ‘pend_restart’ still exists as a readable signal they can be ANDed together with the result ORed into the SDA line just before it is flopped or some other suitable point. Further investigation would still be required to ensure this fix is suitable.

**Impact**

As described by solution 2, an RTL fix implemented as an ECO for this issue is a small tweak and wouldn’t result in any major impact to timing or power. This scenario is already stimulated by our DV, but we do not have an assertion to check this particular protocol violation doesn’t occur, such a check should be added.

**Overall Risk Assessment**
**Tentative Medium**

We need to survey the I2C devices we wish M2.5 silicon to interact with. If all of them are fully usable, even if with reduced performance without RESTARTs and any performance hit isn’t critical this remains medium. The RESTARTs may be usable anyway (as explained) above and we have a software workaround if not.

If the RESTARTs are critical for usage of some devices this becomes a high or critical risk as we are relying on the chance the RESTARTs remain usable. Further work could be done to trial the devices and see how they behave.

**Fix Recommendation**
**Tentative NO FIX**
This is not a critical design issue and arguably not a critical I2C issue as many devices will see no issue with the current behavior (either because RESTART will not be employed or they still see the current behavior as a RESTART) so there is no clear case for an ECO.

**Comment by moidx** [comment]
@cfrantz, @jesultra, @a-will to provide feedback on software impact.

**Comment by a-will** [comment]
As I mentioned in Slack, the abrupt, simultaneous change of sda with scl is due to the restart logic:
https://github.com/lowRISC/opentitan/blob/a68f9f63546d6e96cf41521aabf3b268239f134f/hw/ip/i2c/rtl/i2c_fsm.sv#L501-L505
https://github.com/lowRISC/opentitan/blob/a68f9f63546d6e96cf41521aabf3b268239f134f/hw/ip/i2c/rtl/i2c_fsm.sv#L861-L874

The ClockLow state does not consider that it is going into SetupStart (sda=1, scl=1) when it chooses how to drive SDA. An RTL fix would be to change that:
```diff
       ClockLow : begin
         host_idle_o = 1'b0;
-        sda_d = fmt_byte_i[bit_index];
+        if (pend_restart) begin
+          sda_d = 1'b1;
+        end else begin
+          sda_d = fmt_byte_i[bit_index];
+        end
         scl_d = 1'b0;
       end
```

**Comment by a-will** [comment]
For the "software" perspective, I have nothing to add beyond what Greg pointed out: There are chips that _require_ a repeated start to correctly perform reads (e.g. It's explicitly required for SMBus compatibility with certain protocols).

The question would be whether there is some required capability / compatibility for the shuttle here.

**Comment by GregAC** [comment]
I have put @a-will's fix into a PR: https://github.com/lowRISC/opentitan/pull/18729 so it can be further evaluated. I'm running a reduced nightly regression against it and we should look at the before/after effect with the assertions suggested in https://github.com/lowRISC/opentitan/issues/18722

**Comment by msfschaffner** [comment]
Discussed with @moidx, @a-will and Nuvoton.
Because the fix is small and we found it early, there is a chance that this ECO can be absorbed - but Nuvoton has to check feasibility first.

@GregAC et al, can you prioritize analyzing the regression fallout from #18729 so that we can be sure that no more RTL amendments are needed? Once #18729 is ready, please inform @arnonsha directly so that he can assess feasibility of the ECO.

**Comment by GregAC** [comment]
@hcallahan-lowrisc @marnovandermaas @crteja @engdoreis could you do some FPGA testing with the fix in #18729 ASAP?

**Comment by GregAC** [comment]
I've run a reduced nightly regression on this change (5 seeds per test) and it produce one failure that wasn't seen in other nightlies in the 'i2c_same_csr_outstanding' test

> "../src/lowrisc_ip_i2c_0.1/rtl/i2c_fsm.sv", 1352: tb.dut.i2c_core.u_i2c_fsm.SclOutputGlitch_A: started at 38923674ps failed at 38943674ps
        Offending 'scl_o'

This assert is triggered when the SCL goes high for a single cycle and drops low again, which was suspicious as we've just changed the SCL behavour. I'll look into this more tomorrow. I'm currently running a full nightly regression for I2C.

**Comment by a-will** [comment]
@GregAC If you have the particular seeds used, I could look at it. Out of 50 here, I saw 3 different failures, and none of them were related to the change in the ClockLow state.

However, I _do_ see the `OVRD.TXOVRDEN` CSR getting set in all of them, messing with the actual FSM outputs and causing glitches.

**Comment by johngt** [comment]
See https://github.com/lowRISC/opentitan/pull/18729 for the latest updates.

**Comment by msfschaffner** [comment]
The ECO fix has now been merged.
Please reopen if something else needs to be captured here.
